### PR TITLE
Move TRL loss kernels into TRL

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Moves the source for `trl-lib/trl-losses` into the top-level `kernels/` directory. This keeps the kernel next to TRL without including it in the Python package.

Related to #7244.

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Low Risk**
> Adds isolated kernel build/publish CI and optional accelerator tests; no changes to core TRL training paths in this diff.
> 
> **Overview**
> Adds a top-level **`kernels/`** tree that holds the **`trl-lib/trl-losses`** Hub package source (Triton **`selective_log_softmax_and_entropy`** with forward/backward, optional **`row_mask`** and temperature), plus **`build.toml`**, Nix **`flake`**, and Hub **`README`/`CARD`** templates. The kernels stay **outside** the installable TRL Python package.
> 
> A new **`publish-kernels`** GitHub Action runs on **`main`** when **`kernels/**`** changes (or via **`workflow_dispatch`**): it uses Hugging Face **`kernel-builder`** on HF Jobs to **`check-config`**, **`build-and-upload`** to namespace **`trl-lib`**, and posts to Slack.
> 
> Tests add **`tests/test_kernels.py`** (numerical parity vs PyTorch reference, masked rows, **`torch.compile(fullgraph=True)`**) and **`require_kernels_trust_remote_code`** in **`testing_utils`** so CI only runs Hub kernel tests when **`kernels>=0.14.0`**.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c7155544af91cd4cb04e980384caea45162c928a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: kernels/torch-ext/trl_losses/logprob_entropy.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    log_z_ptr,
+    expected_logit_ptr,
+    logits_batch_stride,
+    logits_row_stride,
+    index_batch_stride,
+    index_row_stride,
+    mask_batch_stride,
+    mask_row_stride,
+    n_cols: tl.constexpr,
+    rows_per_batch: tl.constexpr,
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply kernels/torch-ext/trl_losses/logprob_entropy.py]: Is that correct?

**Review by qgallouedec — COMMENTED** [review]
Super nice!!!
It looks good, just check the bot reviews

**Comment by qgallouedec** [comment]
And we should bump the kernels version in trl pyproject (to the version that introduced `trust_remote_code`, check the ci failure)

**Comment by qgallouedec** [comment]
Thanks, both bot findings look handled. Two follow-ups.

1. Can you drop `transformers[kernels]` / `transformers[hub-kernels]` from the `kernels` and `dev` extras and pin `kernels>=0.14.0` directly? transformers 4.56.2 caps it at `kernels<=0.9`, so on the minimum-versions job `require_kernels_trust_remote_code` skips every run and the new tests never execute. The two have to go for the pin to resolve, `<=0.9` and `>=0.14` can't both hold. Then `require_kernels` is enough and the version guard can go.

2. publish-kernels.yml should report to Slack like every other workflow that runs on main (tests, tests_latest, slow-tests, docker-build). At the top:

```yaml
env:
  CI_SLACK_CHANNEL: ${{ secrets.CI_PUSH_MAIN_CHANNEL }}
```

and as the last step of the publish job:

```yaml
      - name: Post to Slack
        if: ${{ !cancelled() }}
        uses: ./.github/actions/post-slack
        with:
          slack_channel: ${{ env.CI_SLACK_CHANNEL }}
          title: Publish kernels
          slack_token: ${{ secrets.SLACK_CIFEEDBACK_BOT_TOKEN }}
```

File: tests/testing_utils.py [hunk 2]
```diff
@@ -60,6 +60,9 @@
 require_harbor = pytest.mark.skipif(not is_harbor_available(), reason="test requires harbor")
 require_kernels = pytest.mark.skipif(not is_kernels_available(), reason="test requires kernels")
 # `get_kernel(..., trust_remote_code=...)` was added in kernels 0.14.0; older versions don't accept the argument.
+# transformers==4.56.2 (tested by the "minimum versions" CI job) caps `hub-kernels`/`kernels` extras at
+# kernels<=0.9 (huggingface-hub<1.0), and no kernels release since 0.13.0 supports huggingface-hub<1.0, so this
+# can't be resolved by bumping the pyproject.toml floor without dropping support for transformers<5.1.0.
```
Comment (qgallouedec) [inline tests/testing_utils.py:63]: ah right

**Review by qgallouedec — APPROVED** [review]
lgtm! please just fix the ci before merging

**Comment by kashif** [comment]
The direct pin is incompatible with the minimum job: Transformers 4.56.2 requires `huggingface-hub<1.0`, while kernels 0.14 requires `huggingface-hub>=1.10.0`. The metadata guard keeps that job compatible while latest/dev run the kernel tests.

File: kernels/torch-ext/trl_losses/logprob_entropy.py [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+import torch
+import triton
+import triton.language as tl
+
+
+_BLOCK_SIZE = 1024
+
+
+@triton.jit
```
Comment (AmineDiro) [inline kernels/torch-ext/trl_losses/logprob_entropy.py:23]: Nit : 
I wonder if triton warp / block_size autotuning would find some better warp scheduling ?
```
@triton.autotune(configs=[
    triton.Config(kwargs={'BLOCK_SIZE': 128}, num_warps=4),
    triton.Config(kwargs={'BLOCK_SIZE': 1024}, num_warps=8),
  ],
)
```

File: kernels/torch-ext/trl_losses/logprob_entropy.py [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+            tl.store(log_z_ptr + row, 0.0)
+            tl.store(expected_logit_ptr + row, 0.0)
+            return
+    logits_row_ptr = logits_ptr + batch * logits_batch_stride + row_in_batch * logits_row_stride
+    index = tl.load(index_ptr + batch * index_batch_stride + row_in_batch * index_row_stride)
+
+    row_max = -float("inf")
+    denominator = 0.0
+    weighted_sum = 0.0
+    for start in range(0, n_cols, BLOCK_SIZE):
```
Comment (AmineDiro) [inline kernels/torch-ext/trl_losses/logprob_entropy.py:61]: This is super cool ! 

Is thread block handles the full logsumexp so the reduction is probably all SMEM/registers and fast 👏🏼
